pkg/prometheus/operator.go

```diff
@@ -1437,6 +1499,51 @@ func (c *Operator) selectServiceMonitors(p *monitoringv1.Prometheus) (map[string
 	return res, nil
 }
 
+func (c *Operator) selectPodMonitors(p *monitoringv1.Prometheus) (map[string]*monitoringv1.PodMonitor, error) {
+	namespaces := []string{}
+	// Selectors might overlap. Deduplicate them along the keyFunc.
+	res := make(map[string]*monitoringv1.PodMonitor)
+
+	podMonSelector, err := metav1.LabelSelectorAsSelector(p.Spec.PodMonitorSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	// If 'PodMonitorNamespaceSelector' is nil only check own namespace.
+	if p.Spec.PodMonitorNamespaceSelector == nil {
+		namespaces = append(namespaces, p.Namespace)
+	} else {
+		podMonNSSelector, err := metav1.LabelSelectorAsSelector(p.Spec.PodMonitorNamespaceSelector)
+		if err != nil {
+			return nil, err
+		}
+
+		namespaces, err = c.listMatchingNamespaces(podMonNSSelector)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	level.Debug(c.logger).Log("msg", "filtering namespaces to select PodMonitors from", "namespaces", strings.Join(namespaces, ","), "namespace", p.Namespace, "prometheus", p.Name)
+
+	for _, ns := range namespaces {
+		cache.ListAllByNamespace(c.pmonInf.GetIndexer(), ns, podMonSelector, func(obj interface{}) {
+			k, ok := c.keyFunc(obj)
+			if ok {
+				res[k] = obj.(*monitoringv1.PodMonitor)
+			}
+		})
+	}
+
+	podMonitors := []string{}
+	for k := range res {
```

Comment:
Instead of looping again here, we can add the keys to the slice on line 1533, when we are already looping.

Reply:
✅